docs/proposals/petset.md

```diff
@@ -0,0 +1,396 @@
+name is significant, but users are likely to take what they can get. A downside of using unique names
+is that it complicates pre-warming of pods and pod migration - on the other hand, those are also
+advanced use cases that might be better solved by another, more specialized controller (a
+MigratablePetSet).
+
+
+### Managing lifecycle of members
+
+The most difficult aspect of managing a pet set is ensuring that all members see a consistent configuration
+state of the set. Without a strongly consistent view of cluster state, most clustered software is
+vulnerable to split brain. For example, a new set is created with 3 members. If the node containing the
+first member is partitioned from the cluster, it may not observe the other two members, and thus create its
+own cluster of size 1. The other two members do see the first member, so they form a cluster of size 3.
+Both clusters appear to have quorum, which can lead to data loss if not detected.
+
+PetSets should provide basic mechanisms that enable a consistent view of cluster state to be possible,
+and in the future provide more tools to reduce the amount of work necessary to monitor and update that
+state.
+
+The first mechanism is that the PetSet controller blocks creation of new pods until all previous pods
+are reporting a healthy status. The PetSet controller uses the strong serializability of the underyling
+etcd storage to ensure that it acts on a consistent view of the cluster membership (the pods and their)
+status, and serializes the creation of pods based on the health state of other pods. This simplifies
+reasoning about how to initialize a PetSet, but is not sufficient to guarantee split brain does not
+occur.
+
+The second mechanism is having each "pet" use the state of the cluster and transform that into cluster
+configuration or decisions about membership. This is currently implemented using a side car container
+that watches the master (via DNS today, although in the future this may be to endpoints directly) to
+receive an ordered history of events, and then applying those safely to the configuration. Note that
+for this to be safe, the history received must be strongly consistent (must be the same order of
+events from all observers) and the config change must be bounded (an old config version may not
+be allowed to exist forever). For now, this is known as a 'babysitter' (working name) and is intended
+to help identify abstractions that can be provided by the PetSet controller in the future.
+
+
+## Future Evolution
+
+Criteria for advancing to beta:
+
+* Identify common abstraction points from the 'babysitter'
+* One non-database example
+* Make it easy to deploy PetSets without cloud provider storage being required (local storage)
+* Basic rolling-update style upgrades
+
+Criteria for advancing to GA:
+
+* PetSets solve 80% of clustered software configuraton with minimal input from users and are safe from common split brain problems
```

Comment:
I would maybe say that our role is to help do single master leasing (if we can) although I don't think it's a *requirement* for GA.  Maybe have a design for single master leasing and the impacts on the rest of the cluster.

I think organizing arbitrary clustered software is out of scope (and I can add that to the proposal) - any disagreements from the peanut gallery?